### DISCUSSION TITLE
Phantasy Star IV dumps with checksum error

### DISCUSSION THREAD

**Discussion by ghost** [sanni/cartreader#399] [Need help]
Dumped Phantasy Star IV and got a checksum error. Tried loading it in BizHawk and I see the Sega logo and music before it freezes. I tried the FixCheckSum tool, as that worked for me to get Beyond Oasis dump to work, but PSIV does the same thing after checksum fix. How can I dump?

**Comment by sanni** [comment]
You can open the erroneous ROM in an hex editor like Nexus HxD and compare it to a known good dump of the same game. From that comparison you can make assumptions about what did go wrong during the dumping process and then change the dumping function in the source code accordingly. 
It could be that Phantasy Star IV is a special cartridge that needs some extra code to be dumped.

**Comment by ghost** [comment]
That's probably beyond my capabilities, but I'll look into it when I can. 

I am having a similar issue with Beyond Oasis. I thought it was a good dump, as I had tested it to the title screen after initially dumping it. But when I tried to play the dump on Mega ED Pro it freezes when I load a save or start a new game. Same result in BizHawk.

**Comment by Aerik72** [comment]
I am having the same problem dumping Phantasy Star IV. I didn't think it was a problem with my cartridge, as it's in pristine shape, and I cleaned the contacts _throroughly_. A search through discussions led me to this thread.

I don't really have anything new to add, except that it's not just you. :P

**Reply by ghost** [reply]
I don't have a huge collection, but out of my 50 or so carts, PSIV and Beyond Oasis are the only 2 that give me errors. Would love to be able to dump them some day so that I can play my own ROMs in an emulator or on my Everdrive. Cheers.

**Reply by Icesythe7** [reply]
can you send me the dumps, so i can look at them? Icesythe7#9837 is my discord if you want to send there

**Reply by Aerik72** [reply]
Attached here, for you or anybody interested. These are a bunch of my attempted dumps, after cleaning the cartridge contacts.

**Reply by Icesythe7** [reply]
its an exact match till address 0x200000 
![image](https://user-images.githubusercontent.com/13612149/186027081-514759b1-6fc0-4343-8d1b-cf9061c8e47d.png)

**Reply by Aerik72** [reply]
Okay, I could be way off here, I'm out of my element, but check this out: it looks like there are two ROM ICs. Considering the memory address it gets to, is it possible it's reading the first ROM IC, but not the second, so it's only getting half the data?

![20220823_000924](https://user-images.githubusercontent.com/88597229/186075409-b7dba77f-4196-4a03-9916-c329802bdfcb.jpg)
![20220823_000931](https://user-images.githubusercontent.com/88597229/186075415-03ef9a29-7892-41b1-913c-dda2f13e3bd8.jpg)

**Reply by ghost** [reply]
Beyond Oasis has a similarly large board with multiple ICs, IIRC.

**Reply by Icesythe7** [reply]
according to my reading someone would have to implement this
![image](https://user-images.githubusercontent.com/13612149/186348524-fed1a6a7-7add-4c8a-9ba9-b254b6848a25.png)
which look like sanni may have this already but maybe its something else special on these games or not working properly? I dont have a game like this to test sadly and am already deep into a gameboy rewrite or I'd look into it further for you.

**Comment by sanni** [comment]
Could be like this image I found on google, one 74HC74 and one 74HC139:

![unnamed](https://user-images.githubusercontent.com/19169494/186373993-aac47c9a-fe53-48b4-a203-04331ff69544.png)
![decodificador-demultiplexor-de-2-a-4-lineas-74hc139-3491862534](https://user-images.githubusercontent.com/19169494/186375712-a252161d-0cb5-40ab-8484-117aca90bdd0.jpg)

To enable ROM the output 3 (top row) of the 139 needs to be LOW. For that AB needs to be HIGH. So the bottom output 3 needs to be HIGH, which is the case as long as not both A and B (bottom row) are HIGH. Which means the function enableSram_MD(0) needs to be called before dumping the ROM to set the 74HC74 output to LOW.

Can you try this changed MD.ino (just replace the one in V9.5_Portable\Arduino IDE Portable\portable\sketchbook\Cart_Reader\ and upload)

[MD.zip](https://github.com/sanni/cartreader/files/9415089/MD.zip)

**Reply by Aerik72** [reply]
It works!! :D :D The ROM is good, boots up in an emulator and everything.

At least, it does for Phantasy Star IV. I don't have Beyond Oasis, so somebody else will have to check.

Thanks, Sanni!!

![20220824_110904](https://user-images.githubusercontent.com/88597229/186469279-5421741b-46c8-43ae-86b8-ef7aa4c11578.jpg)

**Reply by Icesythe7** [reply]
Print a frame son!

**Reply by Aerik72** [reply]
Oh, uhhhhh.... I actually don't have a 3D printer. :P

**Reply by ghost** [reply]
> Can you try this changed MD.ino (just replace the one in V9.5_Portable\Arduino IDE Portable\portable\sketchbook\Cart_Reader\ and upload)
> 
> [MD.zip](https://github.com/sanni/cartreader/files/9415089/MD.zip)

Where do I put that? How do I get to that folder?

**Reply by sanni** [reply]
Download the [9.5 portable release](https://github.com/sanni/cartreader/releases/tag/V9.5) and the [updated code from Github](https://github.com/sanni/cartreader/archive/refs/heads/master.zip) and then copy and overwrite the Cart_Reader folder found in V9.5_Portable\Arduino IDE Portable\portable\sketchbook\ with the updated files and [flash the Arduino](https://github.com/sanni/cartreader/wiki/How-to-flash-the-Arduino).

**Reply by ghost** [reply]
Dumped both PSIV and Beyond Oasis and tested in BizHawk. Both work! Will test on my EverDrive next.

**Reply by PsyK0p4T** [reply]
Did you get the expected crc32 for both ? It's more important than running them in an emulator/flashcart ^^

**Reply by ghost** [reply]
I don't know. Is that what would have displayed when I dumped them? I would have to dump them again.
